### PR TITLE
Accelerate Instant-NGP inference

### PR DESCRIPTION
**Pull request by Linyou**
This PR enhances Nerfacc's Instant-NGP inference performance by implementing the following API changes:

1. The `traverse_grids` function has been modified to support both train and test modes.
2. A new function called `mark_invisible_cells` has been added to the `occ_grid` module in order to prevent rendering artifacts in unseen spaces.

### CODE REVIEW

File: nerfacc/cuda/csrc/grid.cu [hunk 1]
```diff
@@ -237,6 +237,177 @@ __global__ void traverse_grids_kernel(
...
+    // parallelize over rays
+    for (int32_t tid = blockIdx.x * blockDim.x + threadIdx.x; tid < n_rays; tid += blockDim.x * gridDim.x)
+    {
+
+        num_steps += tid;
+        continuous_resume += tid;
+        t_starts += tid * N_samples;
+        t_ends += tid * N_samples;
+        valid_mask += tid * N_samples;
+        ray_indices += tid * N_samples;
```
Comment (liruilong940607) [inline nerfacc/cuda/csrc/grid.cu:279]: Can't use `+=` anymore because of the forloop:

    for (int32_t tid = blockIdx.x * blockDim.x + threadIdx.x; tid < n_rays; tid += blockDim.x * gridDim.x)

It would be wrong if `+=` is executed twice within the forloop.

use something like `valid_mask[tid * Nsamples]` to read/write the data

**Comment by liruilong940607** [comment]
Thanks for implementing this! It is pretty nice to have that support

**Comment by liruilong940607** [comment]
On the high level, the two ways of ray marching are pretty similar to each other: The "train" way of marching is to take N rays and march all the steps for each ray. The "test" ways of marching is to take all rays and march N steps for each ray (and iterative).

I feel it should be not that hard to unify the API (as well as the implementation) for the two. 

To be more concrete, implementation differences between the two are:
- "test" way needs to take in an argument "max_per_ray_samples", for which the "train" way could simply set it to "inf".
- "test" way would want to pre-compute the "{t_sorted, t_indices, hits}" so that they are not computed multiple times. As you have already done, we should do this in python instead of C.  So that we can make these three as optional arguments for the API that allows for passing in precomputed values.
- "test" way needs to pass in a mask with shape (n_rays,) to skip rays. (e.g., the alive_indices you were using). "train" way can have an all ones mask.

So maybe we can unify them into the same "traverse_grid" function, with extra arguments (max_per_ray_samples=inf, masks=None, t_sorted=None, t_indices=None, hits=None). And for "traverse_grid_test", your can just call that function with an updated "near_planes" at every iteration of marching.

In this case, I think it makes sense to let the CUDA kernel return an extra tensor (n_rays,) that indicates the termination distance during grid traversal, which is essentially the "near_planes" for the next iteration of "traverse_grid_test". (the `near_planes` you are returning has a confusing name, which I think it should be `termination_planes` or something like that.

**Comment by Linyou** [comment]
Sound good! I think we could unify the API using extra arguments "(max_per_ray_samples=inf, masks=None, t_sorted=None, t_indices=None, hits=None)". Nice ideal BTW! 

We also need to unify the return value, I suggest using the data structure defined in "data_spect.h" to store `t_start` and `t_end`, instead of creating torch::Tensor directly as I am currently doing. It may be helpful to add new methods for allocating memory in "data_spect.h" specifically for `t_start` and `t_end", since they are pre-allocated in the "test" way. What do you think?

As for `near_planes`, I already tweaked the code so we don't need to return it, we can just update it inside the "traverse_grid" kernel.

**Comment by liruilong940607** [comment]
> We also need to unify the return values. I suggest using the data structure defined in "data_spect.h" to store t_start and t_end, instead of creating torch::Tensor directly as I am currently doing. It may be helpful to add new methods for allocating memory in "data_spect.h" specifically for t_start and `t_end", since they are pre-allocated in the "test" way. What do you think?

I think you can use the [RaySegmentsSpec](https://github.com/KAIR-BAIR/nerfacc/blob/master/nerfacc/cuda/csrc/include/data_spec.hpp#L44) just like what is being used in the `traverse_grid` function. And you can get `t_starts` and `t_ends` by:

https://github.com/KAIR-BAIR/nerfacc/blob/8340e19daad4bafe24125150a8c56161838086fa/tests/test_grid.py#L60-L61

> As for near_planes, I already tweaked the code so we don't need to return it, we can just update it inside the "traverse_grid" kernel.

Do you mean that you inplace change the value of it? I would suggest against doing inplace modification as it is not quite user-friendly.

File: nerfacc/cuda/csrc/grid.cu [hunk 2]
```diff
@@ -305,80 +356,127 @@ std::vector<RaySegmentsSpec> traverse_grids(
...
+            intervals.chunk_cnts = torch::full({n_rays}, -1, rays_o.options().dtype(torch::kLong));
+            intervals.chunk_starts = torch::arange(
+                0, total_steps * 2, traverse_steps_limit * 2, rays_o.options().dtype(torch::kLong));
+            intervals.memalloc_data(total_steps * 2, true, true);
+        }
+        if (compute_samples) {
+            samples.chunk_cnts = torch::full({n_rays}, -1, rays_o.options().dtype(torch::kLong));
+            samples.chunk_starts = torch::arange(
+                0, total_steps, traverse_steps_limit, rays_o.options().dtype(torch::kLong));
+            samples.memalloc_data(total_steps, false, true, true);
```
Comment (Linyou) [inline nerfacc/cuda/csrc/grid.cu:378]: When in test mode, if we have the full `n_rays` each time (which will be contained within a while loop), memory allocation will become excessive due to the increase of `traverse_steps_limit` within the loop. However, if we only pass alive rays and not the full `n_rays`, then the `ray_mask` becomes redundant as it only indicates whether or not a ray is alive.

Comment (liruilong940607) [reply nerfacc/cuda/csrc/grid.cu:378]: Good points. I didn't think of that. 

So how about we do this -- We get rid of the `rays_mask`, and consider all rays that being passed to C are alive rays. And when we use the `traverse_grids` function in python, we pass it only the alive rays by masking out others. Something like:

```
travers_grids(rays_o[masks], rays_d[masks] ...)
```

File: nerfacc/cuda/csrc/grid.cu [hunk 3]
```diff
@@ -207,13 +250,15 @@ __global__ void traverse_grids_kernel(
...
+                                samples.is_valid[idx] = true;
                             }
-                            n_samples++;
                         }
 
+                        n_samples++;
                         continuous = true;
                         t_last = t_next;
                         if (t_next >= t_traverse) break;
+                        if (traverse_steps_limit > 0 && n_samples >= traverse_steps_limit) break;
```
[the remark this answers is not in the document]
Comment (liruilong940607) [reply nerfacc/cuda/csrc/grid.cu:261]: That's true. Will fix!

[the remark this answers is not in the document]
Comment (liruilong940607) [reply nerfacc/cuda/csrc/grid.cu]: resolved in the latest commit

File: nerfacc/cuda/csrc/grid.cu (hunk 2 above)
Comment (liruilong940607) [reply nerfacc/cuda/csrc/grid.cu]: resolved in the latest commit

**Comment by liruilong940607** [comment]
@Linyou The latest commit should resolve the memory concerns we had before. The test is also updated to match with the actual use case. Lmk what do you think.

**Comment by Linyou** [comment]
Thanks! I believe that the current API design is now highly usable for test mode rendering, thanks to the latest commit. 

BTW, after this PR is merged, I will create a new one for ngp test mode rendering in the examples.

**Comment by liruilong940607** [comment]
@Linyou I also did some cleanups for `mark_invisible_cells()` and changed the API a tiny bit (the `K`). Now I'm happy to merge it if you think it's all good.

File: nerfacc/estimators/occ_grid.py [hunk 4]
```diff
@@ -258,6 +258,80 @@ def update_every_n_steps(
...
+                # a valid cell should be visible by at least one camera and not too close to any camera
+                valid_mask = (count > 0) & (~too_near_to_any_cam)
+
+                cell_ids_base = lvl * self.cells_per_lvl
+                self.occs[cell_ids_base + indices_chunk] = torch.where(
+                    valid_mask, 0.0, -1.0
+                )
+
     @torch.no_grad()
     def _get_all_cells(self) -> List[Tensor]:
```
Comment (Linyou) [inline nerfacc/estimators/occ_grid.py:336]: I think we also need to filter out the cells with -1 density in here. Otherwise, the `mark_invisible_cells` will become ineffective.

File: nerfacc/estimators/occ_grid.py [hunk 5]
```diff
@@ -271,6 +345,9 @@ def _sample_uniform_and_occupied_cells(self, n: int) -> List[Tensor]:
             uniform_indices = torch.randint(
                 self.cells_per_lvl, (n,), device=self.device
             )
+            # filter out the cells with -1 density (non-visible to any camera)
+            cell_ids = lvl * self.cells_per_lvl + uniform_indices
+            uniform_indices = uniform_indices[self.occs[cell_ids] >= 0.0]
```
Comment (Linyou) [inline nerfacc/estimators/occ_grid.py:350]: Yes, we should also include this in `_get_all_cells` since we update all cells during the warm-up stage.

File: nerfacc/estimators/occ_grid.py [hunk 6]
```diff
@@ -258,6 +258,80 @@ def update_every_n_steps(
...
+
+        Args:
+            K: Camera intrinsics of shape (N, 3, 3) or (1, 3, 3).
+            c2w: Camera to world poses of shape (N, 3, 4) or (N, 4, 4).
+            width: Image width in pixels
+            height: Image height in pixels
+            near_plane: Near plane distance
+            chunk: The chunk size to split the cells (to avoid OOM)
+        """
+        assert K.dim() == 3 and K.shape[1:] == (3, 3)
```
Comment (Linyou) [inline nerfacc/estimators/occ_grid.py:285]: Should we ensure that the `c2w` and `K` share the same value for `N`?

File: nerfacc/estimators/occ_grid.py (hunk 4 above)
Comment (liruilong940607) [reply nerfacc/estimators/occ_grid.py:336]: Yeah nice catch!

File: nerfacc/estimators/occ_grid.py (hunk 6 above)
Comment (liruilong940607) [reply nerfacc/estimators/occ_grid.py:285]: Will add

**Comment by liruilong940607** [comment]
Comments addressed. Ready to Go? @Linyou

**Comment by Linyou** [comment]
@liruilong940607 Yeah! All good!

**Comment by liruilong940607** [comment]
Thanks for the patience!! Shipped!
